### ISSUE TITLE
component push throws error of waited 4m0s but couldn't find running pod matching selector

### ISSUE THREAD

**Issue by prietyc123** [openshift/odo#2877]
/kind flake

## What versions of software are you using?

**Operating System:**
All Supported

**Output of `odo version`:**
master

## How did you run odo exactly?
```odo push --context context``` on OpenShift ci.

## Actual behavior
Throwing error as: 
```
[odo] Please use `odo push` command to create the component with source deployed
Running odo with args [odo push --context /tmp/947847413]
[odo] Validation
[odo]  •  Checking component  ...
[odo] 
 ✓  Checking component [60ms]
[odo] 
[odo] Configuration changes
[odo]  ✓  Initializing component
[odo]  •  Creating component  ...
[odo] 
 ✓  Creating component [200ms]
[odo] 
[odo] Pushing to component dotnet-app of type local
[odo]  •  Checking files for pushing  ...
[odo] 
 ✓  Checking files for pushing [553410ns]
[odo]  •  Waiting for component to start  ...
[odo]  ✗  Waiting for component to start [4m]
[odo]  ✗  waited 4m0s but couldn't find running pod matching selector: 'deploymentconfig=dotnet-app-app'
Deleting project: bkbwthrcvq
Running odo with args [odo project delete bkbwthrcvq -f]
```

## Expected behavior
It should push the component successfully into the deployment.

## Any logs, error output, etc?

```
[odo] Please use `odo push` command to create the component with source deployed
Running odo with args [odo push --context /tmp/947847413]
[odo] Validation
[odo]  •  Checking component  ...
[odo] 
 ✓  Checking component [60ms]
[odo] 
[odo] Configuration changes
[odo]  ✓  Initializing component
[odo]  •  Creating component  ...
[odo] 
 ✓  Creating component [200ms]
[odo] 
[odo] Pushing to component dotnet-app of type local
[odo]  •  Checking files for pushing  ...
[odo] 
 ✓  Checking files for pushing [553410ns]
[odo]  •  Waiting for component to start  ...
[odo]  ✗  Waiting for component to start [4m]
[odo]  ✗  waited 4m0s but couldn't find running pod matching selector: 'deploymentconfig=dotnet-app-app'
Deleting project: bkbwthrcvq
Running odo with args [odo project delete bkbwthrcvq -f]
```
For more details: https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_odo/2875/pull-ci-openshift-odo-master-v4.1-integration-e2e-benchmark/1778#1:build-log.txt%3A710

**Comment by amitkrout** [comment]
similar error with little bit twist - https://github.com/openshift/odo/pull/2942#issuecomment-617111589

**Comment by amitkrout** [comment]
@mik-dass This issue appears more frequently than before when the test node was 2. I am also assigning you to this issue along with me as you have fixed similar kind of issue before.

**Comment by mik-dass** [comment]
> @mik-dass This issue appears more frequently than before when the test node was 2. I am also assigning you to this issue along with me as you have fixed similar kind of issue before.

Let's reduce the number of nodes back to 2.

**Comment by prietyc123** [comment]
I agree to 

> > @mik-dass This issue appears more frequently than before when the test node was 2. I am also assigning you to this issue along with me as you have fixed similar kind of issue before.
> 
> Let's reduce the number of nodes back to 2.

Not a bad though I suspect the failure might be due to less resources in travis CI. Is there any way to increase the resouces in travis while doing ```oc cluster up```? cc_ @amitkrout

**Comment by prietyc123** [comment]
one more hit in https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_odo/2965/pull-ci-openshift-odo-master-v4.2-integration-e2e-benchmark/2313#1:build-log.txt%3A378

**Comment by prietyc123** [comment]
Hitting this issue more frequently when running tests with ```xenial``` distribution of travis CI https://travis-ci.com/github/openshift/odo/jobs/324857465#L473 . 

```Xenial``` distribution of travis is required to run latest kubernetes cluster. On the other hand we can use older version of minikube but ```odo push``` does not support the older version of minikube, which has been elaborated in issue : https://github.com/openshift/odo/issues/2928 . 

However there are some other consequences also of using older version of minikube like lag of latest feature implementation and I am suspecting this could be one of the reason that ```odo push``` is not supporting on older version of minikube. IMO we should move towards latest version minikube and hence ```xenial``` distribution of travis CI could be one of the solution.

Right now we are running our test on travis CI with ```trusty``` distribution, ubuntu version 14.04. And running the latest minikube needs ```systemd``` which was added in ubuntu 16.04, therefore we need to bump the ```ubuntu``` version to 16.04+ using ```xenial``` distribution of travis CI.

@kadel @girishramnani WDYT?

**Comment by prietyc123** [comment]
> > > @mik-dass This issue appears more frequently than before when the test node was 2. I am also assigning you to this issue along with me as you have fixed similar kind of issue before.
> > 
> > 
> > Let's reduce the number of nodes back to 2.
> 
> Not a bad though I suspect the failure might be due to less resources in travis CI. Is there any way to increase the resources in travis while doing `oc cluster up`? cc_ @amitkrout

I have raised a ticket asking for more resources on travis CI. Let see what they are replying on that. I will update the same once the reply

**Comment by prietyc123** [comment]
> I have raised a ticket asking for more resources on travis CI. Let see what they are replying on that. I will update the same once the reply

Got the response from Travis CI team ![image](https://user-images.githubusercontent.com/51620925/80594799-4a2a1280-8a41-11ea-865a-52cb70fe8eab.png)

The provided default memory 7.5 is enough for running 4 component push in parallel. So i think we should check component push failure from odo end. WDYT mrinal ?

**Comment by mik-dass** [comment]
> The provided default memory 7.5 is enough for running 4 component push in parallel. So i think we should check component push failure from odo end. WDYT @mik-dass ?

But decreasing the test nodes to 2 has indeed reduced the amount of this failure. Also 7.5 may not be enough as we are running minishift in the background in most of our tests. Also the pod initialization step can consume a lot of the time too. I would suggest increasing the push timeout value by `odo preference set pushtimeout <some higher value>  -f` in most of our tests, increasing the number of test nodes to 4 and verifying.

**Comment by prietyc123** [comment]
Hitting same flake on openshift release repo https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/8633/rehearse-8633-pull-ci-openshift-odo-master-v4.2-integration-e2e/2#1:build-log.txt%3A426

**Comment by amitkrout** [comment]
> > The provided default memory 7.5 is enough for running 4 component push in parallel. So i think we should check component push failure from odo end. WDYT @mik-dass ?
> 
> But decreasing the test nodes to 2 has indeed reduced the amount of this failure. Also 7.5 may not be enough as we are running a cluster in the background in most of our tests which can be a expensive operation too. Also the pod initialization step can consume a lot of the time. I would suggest increasing the push timeout value by `odo preference set pushtimeout <some higher value> -f` in most of our tests, increasing the number of test nodes to 4 and verifying.

@mik-dass may be you are right, however iirc even on single test node we had the similar issue. Anyway we can try your suggestion to narrow down the reason for failure.

**Comment by amitkrout** [comment]
@prietyc123 Can you please apply @mik-dass suggestion in one of your pr you mentioned in the comment https://github.com/openshift/odo/issues/2877#issuecomment-621117106. 

You just need to overwrite the ```pushtimeout``` through the global config. Lets make the wait time twice the actual i.e. 8 minute.

**Comment by prietyc123** [comment]
> @prietyc123 Can you please apply @mik-dass suggestion in one of your pr you mentioned in the comment [#2877 (comment)](https://github.com/openshift/odo/issues/2877#issuecomment-621117106).
> 
> You just need to overwrite the `pushtimeout` through the global config. Lets make the wait time twice the actual i.e. 8 minute.

Sure I will definitely try it out and update the result.

**Comment by prietyc123** [comment]
> > The provided default memory 7.5 is enough for running 4 component push in parallel. So i think we should check component push failure from odo end. WDYT @mik-dass ?
> 
> But decreasing the test nodes to 2 has indeed reduced the amount of this failure. Also 7.5 may not be enough as we are running a cluster in the background in most of our tests which can be a expensive operation too. Also the pod initialization step can consume a lot of the time. I would suggest increasing the push timeout value by `odo preference set pushtimeout <some higher value> -f` in most of our tests, increasing the number of test nodes to 4 and verifying.

@mik-dass I have set the pushtimeout to 8 min but still getting the same failure.
```
Running odo with args [odo push --context /tmp/280910682]
[odo] Validation
[odo]  •  Checking component  ...
 ✓  Checking component [11ms]
[odo] 
[odo] Configuration changes
[odo]  ✓  Initializing component
[odo]  •  Creating component  ...
 ✓  Creating component [122ms]
[odo] 
[odo] Applying URL changes
[odo]  ✓  URL warfile: http://warfile-app-yweencucjw.127.0.0.1.nip.io created
[odo] 
[odo] Pushing to component javaee-war-test of type binary
[odo]  •  Checking files for pushing  ...
 ✓  Checking files for pushing [302877ns]
[odo]  •  Waiting for component to start  ...
 ✗  waited 8m0s but was unable to find a running pod matching selector: 'deploymentconfig=javaee-war-test-app'
[odo] For more information to help determine the cause of the error, re-run with '-v'.
[odo] See below for a list of failed events that occured more than 5 times during deployment:
[odo] 
[odo]  NAME                                          COUNT  REASON  MESSAGE                 
[odo] 
[odo]  javaee-war-test-app-1-cljpt.160a8760de1a2fe9  20     Failed  Error: ImagePullBackOff 
[odo] 
[odo] 
[odo]  ✗  Waiting for component to start [8m] [WARNING x20: Failed]
Deleting project: yweencucjw
Running odo with args [odo project delete yweencucjw -f]
[odo] This project contains the following applications, which will be deleted
[odo] Application app
[odo] This application has following components that will be deleted
[odo] component named javaee-war-test
[odo] This component has following urls that will be deleted with component
[odo] URL named warfile with host warfile-app-yweencucjw.127.0.0.1.nip.io having protocol http at port 8080
[odo] No services / could not get services
[odo]  •  Deleting project yweencucjw  ...
 ✓  Deleting project yweencucjw [6s]
[odo]  ✓  Deleted project : yweencucjw
Deleting dir: /tmp/280910682
• Failure [488.196 seconds]
odo java e2e tests
/home/travis/gopath/src/github.com/openshift/odo/tests/e2escenarios/e2e_java_test.go:13
  odo component creation
  /home/travis/gopath/src/github.com/openshift/odo/tests/e2escenarios/e2e_java_test.go:42
    Should be able to deploy a .war file using wildfly [It]
    /home/travis/gopath/src/github.com/openshift/odo/tests/e2escenarios/e2e_java_test.go:59
    No future change is possible.  Bailing out early after 480.422s.
    Running odo with args [odo push --context /tmp/280910682]
    Expected
        <int>: 1
    to match exit code:
        <int>: 0
```
More details : https://travis-ci.com/github/openshift/odo/jobs/325397186#L1861

**Comment by mik-dass** [comment]
> @mik-dass I have set the pushtimeout to 8 min but still getting the same failure.

There seems to be some issue with the network on travis and most probably the ImagePullBackOff happened because of that

https://travis-ci.com/github/openshift/odo/jobs/325397186#L1693
https://travis-ci.com/github/openshift/odo/jobs/325397186#L625

Also TBH I haven't seen this error on most PRs since we switched back to 2 nodes. Even on 4 nodes it happened in 2-4 test scripts.

But for your PR https://github.com/openshift/odo/pull/2913 it's happening for all the test scripts. In fact all the jobs on Travis, which run on `oc cluster up`, are failing with almost the same error (either image pull back off or http connection error) https://travis-ci.com/github/openshift/odo/builds/162887080. I would advise you to 

- revert all the changes in your PR and test on Travis
- switch the distro to xenial and test
- implement all the changes step by step in your PR and test on Travis and report back

**Comment by mik-dass** [comment]
Here's an recent incident on CI https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_odo/3054/pull-ci-openshift-odo-master-v4.2-integration-e2e-benchmark/2549#1:build-log.txt%3A382

**Comment by prietyc123** [comment]
one more occurence on CI https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_odo/2913/pull-ci-openshift-odo-master-v4.2-integration-e2e/78#1:build-log.txt%3A424

**Comment by amitkrout** [comment]
Hitting it on OpenShift CI too - https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/8707/rehearse-8707-periodic-ci-openshift-odo-master-v4.2-integration-e2e-periodic/3#1:build-log.txt%3A482

@mik-dass I think its time to revisit the issue

**Comment by mik-dass** [comment]
> @mik-dass I think its time to revisit the issue

let's increase the pushtimeout and check

**Comment by amitkrout** [comment]
> > @mik-dass I think its time to revisit the issue
> 
> let's increase the pushtimeout and check

@mik-dass This is not the push time out failure, its actually component build timeout failure. I was following a bugzila where the user had a similar issue on odo 1.1.3. Here i am pasting description and the workaround provided by the user

User says:
```
I can confirm that the problem is the build timeout in odo.

I have modified the odo client, cloning the github repo, modified the file pkg/occlient/occlient.go and increase the value from the variable OcBuildTimeout in line 99 to 15 (default was 5), and the build finished successfully, injecting the resulting image name into the deploymentconfig.

diff --git a/pkg/occlient/occlient.go b/pkg/occlient/occlient.go
index 6e59dd646..637016acc 100644
--- a/pkg/occlient/occlient.go
+++ b/pkg/occlient/occlient.go
@@ -95,8 +95,8 @@ type CreateArgs struct {

 const (
        failedEventCount                = 5
-       OcUpdateTimeout                 = 5 * time.Minute
-       OcBuildTimeout                  = 5 * time.Minute
+       OcUpdateTimeout                 = 15 * time.Minute
+       OcBuildTimeout                  = 15 * time.Minute
        OpenShiftNameSpace              = "openshift"
        waitForComponentDeletionTimeout = 120 * time.Second

I believe the timeout in the default value is reached due to the amount of maven packages the build needs to download, and since our connection is not the fastest, the build reach this timeout and exits.

Perhaps you can add an option in the future where the user can define the timeout value for the build in slow or low processing environments.
```

Now its clear that the issue not with the current odo backend code rather its an environment specific issue (I mean slow env or slow network), however we should expose a key that would overwrite the default build time out in case of slower environment.

**Comment by mik-dass** [comment]
> Now its clear that the issue not with the current odo backend code rather its an environment specific issue (I mean slow env or slow network), however we should expose a key that would overwrite the default build time out in case of slower environment.

OK let's make a issue about increasing the build timeout

**Comment by amitkrout** [comment]
> > Now its clear that the issue not with the current odo backend code rather its an environment specific issue (I mean slow env or slow network), however we should expose a key that would overwrite the default build time out in case of slower environment.
> 
> OK let's make a issue about increasing the build timeout

https://github.com/openshift/odo/issues/3110

**Comment by mik-dass** [comment]
> For more details: https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_odo/2875/pull-ci-openshift-odo-master-v4.1-integration-e2e-benchmark/1778#1:build-log.txt%3A710

Looking at the above log and the others, they seem to be local components.
```
Running odo with args [odo create dotnet:2.0 dotnet-app --project bkbwthrcvq --context /tmp/947847413]
[odo] Validation
[odo] Warning: dotnet:2.0 is not fully supported by odo, and it is not guaranteed to work
[odo]  •  Validating component  ...
[odo] 
 ✓  Validating component [40ms]
[odo] 
[odo] Please use `odo push` command to create the component with source deployed
Running odo with args [odo push --context /tmp/947847413]
[odo] Validation
[odo]  •  Checking component  ...
```
Build timeout is only used for git components. So https://github.com/openshift/odo/issues/3110 won't resolve this issue.

**Comment by prietyc123** [comment]
Hitting on OpenShift CI https://deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gcs/origin-ci-test/pr-logs/pull/openshift_odo/3252/pull-ci-openshift-odo-master-v4.2-integration-e2e/710#1:build-log.txt%3A1511

**Comment by prietyc123** [comment]
> Hitting it on OpenShift CI too - https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/8707/rehearse-8707-periodic-ci-openshift-odo-master-v4.2-integration-e2e-periodic/3#1:build-log.txt%3A482
> 
> @mik-dass I think its time to revisit the issue

Similar type of Failure https://deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gcs/origin-ci-test/pr-logs/pull/openshift_odo/3252/pull-ci-openshift-odo-master-v4.2-integration-e2e/710#1:build-log.txt%3A1511

**Comment by amitkrout** [comment]
Hit the same issue in periodic job too - https://deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gcs/origin-ci-test/logs/periodic-ci-openshift-odo-master-v4.3-integration-e2e-periodic/1274764207098695680#1:build-log.txt%3A592

**Comment by amitkrout** [comment]
one more - https://deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gcs/origin-ci-test/logs/periodic-ci-openshift-odo-master-v4.2-integration-e2e-periodic/1275670220752556032#1:build-log.txt%3A441

**Comment by amitkrout** [comment]
> > For more details: https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_odo/2875/pull-ci-openshift-odo-master-v4.1-integration-e2e-benchmark/1778#1:build-log.txt%3A710
> 
> Looking at the above log and the others, they seem to be local components.
> 
> ```
> Running odo with args [odo create dotnet:2.0 dotnet-app --project bkbwthrcvq --context /tmp/947847413]
> [odo] Validation
> [odo] Warning: dotnet:2.0 is not fully supported by odo, and it is not guaranteed to work
> [odo]  •  Validating component  ...
> [odo] 
>  ✓  Validating component [40ms]
> [odo] 
> [odo] Please use `odo push` command to create the component with source deployed
> Running odo with args [odo push --context /tmp/947847413]
> [odo] Validation
> [odo]  •  Checking component  ...
> ```
> 
> Build timeout is only used for git components. So #3110 won't resolve this issue.

@mik-dass 
ohh, is it ?
Can we then plan for a separate issue for local component ? or want to fix in the same pr #3110

**Comment by mik-dass** [comment]
> Can we then plan for a separate issue for local component ? or want to fix in the same pr #3110

The PR build timeout PR will only close https://github.com/openshift/odo/issues/3110. This issue should remain open.

**Comment by amitkrout** [comment]
once a while we are hitting this issue on prow CI. So i am adding it to the current sprint planning to see some progress on this flake.

**Comment by kadel** [comment]
> once a while we are hitting this issue on prow CI. So i am adding it to the current sprint planning to see some progress on this flake.

@amitkrout 
there have been some additional logs added recently to help debug this, can you link to logs or jobs where this happened more recently?

**Comment by sarveshtamba** [comment]
@prietyc123 any progress on this ? Seeing this on one of our clusters frequently again.

**Comment by dharmit** [comment]
I'm reducing the priority on this one since we're already working on #3256 and decided to focus on that one first.

/remove-priority high
/priority medium
